svelte-kit-scss/src/routes/(authenticated)/[username]/[repo]/[issueSearchType=issue_search_type]/+page.server.ts

```diff
@@ -0,0 +1,58 @@
+import type { PageServerLoad } from './$types';
+import {
+  IssueSearchQueryState,
+  IssuesSearchQuerySort,
+  type IssueSearchQueryFilter,
+} from '$lib/constants/issues-search-query-filters';
+import {
+  buildIssueSearchQuery,
+  ensureIssuesSearchQueryFilterValue,
+  parseIssuesSearchQueryAllowed,
+} from '$lib/helpers/issues-search-query-builder';
+import { IssuesSearchService } from '$lib/services';
+import { IssueSearchPageTypeFiltersMap, type IssueSearchTypePage } from '$lib/constants/matchers';
+
+const DEFAULT_PER_PAGE = 25;
+const PAGE_SEARCH_PARAM_QUERY = 'q';
+const PAGE_SEARCH_PARAM_PAGE = 'page';
+
+export const load: PageServerLoad = async ({ fetch, params, url: { searchParams } }) => {
+  const service = new IssuesSearchService(fetch, DEFAULT_PER_PAGE);
+  const { username, repo, issueSearchType } = params;
+
+  const defaultFilters: IssueSearchQueryFilter[] = [
+    IssueSearchPageTypeFiltersMap[issueSearchType as IssueSearchTypePage],
+    IssueSearchQueryState.Open,
+    IssuesSearchQuerySort.Newest,
+  ];
+
+  const currentSearchQuery = searchParams.get(PAGE_SEARCH_PARAM_QUERY);
+  const currentPageString = searchParams.get(PAGE_SEARCH_PARAM_PAGE);
+  const currentPage = currentPageString ? Number(currentPageString) : null;
+
+  const allowedSearchQueryFilters =
+    parseIssuesSearchQueryAllowed(currentSearchQuery) || defaultFilters;
+  const searchQuery = buildIssueSearchQuery(allowedSearchQueryFilters, `${username}/${repo}`);
+
+  const issues = await service.getIssues(searchQuery, { page: currentPage });
```

Comment:
If you remove these `await`s it will be better.

As is:

- first it makes a request for the issues
- then, when that is resolved, it makes a fetch for the open count,
- then, when that is resolved, it makes a fetch for the closed count

Now luckily, sveltekit does this nice thing where it awaits the top level properties returned from load, so you only really have to remove the awaits here and everything should work, and not have that cascading wait

Reply:
It does not.

Though the SO answer linked is mostly correct.
You have to think about where the code "suspends", which is where the `await` keyword is exectuted.

In the SO answer linked it describes initiating 3 promises eagerly, then awaiting all three, after they have already begun. So the time there is basically waiting for the longest one.

In this code, the code in this file, the await happens at the creation of each of the promises.  So it "suspends" execution of the code until the first promise is resolved, then the code continues, then it "suspends" execution again until the second resolves, and so on.

If you'd like to jump on a call to discuss further let me know.

Reply:
If this helps run these in your console:

```js
function wait(seconds) { return new Promise(res => setTimeout(res, seconds*1000)) }

async function example() {
    console.log('Begin')
    await wait(1)
    console.log('Got one!')
    await wait(1)
    console.log('...and another!')
    await wait(1)
    console.log('...got all 3!')
}

async function example2() {
    console.log('Begin')
    const w1 = wait(1)
    const w2 = wait(1)
    const w3 = wait(1)
    await w1;
    console.log('Got one!')
    await w2;
    console.log('...and another!')
    await w3;
    console.log('...got all 3!')
}
```